drivers/storage/port/scsiport/fdo.c

```diff
@@ -705,6 +705,98 @@ FdoHandleDeviceRelations(
     return IoCallDriver(PortExtension->Common.LowerDevice, Irp);
 }
 
+static
+NTSTATUS
+NTAPI
+ForwardIrpAndWaitCompletion(
+    _In_ PDEVICE_OBJECT DeviceObject,
+    _In_ PIRP Irp,
+    _In_ PVOID Context)
+{
+    if (Irp->PendingReturned)
+        KeSetEvent((PKEVENT)Context, IO_NO_INCREMENT, FALSE);
+    return STATUS_MORE_PROCESSING_REQUIRED;
+}
+
+static
+NTSTATUS
+ForwardIrpAndWait(
+    _In_ PDEVICE_OBJECT LowerDevice,
+    _In_ PIRP Irp)
+{
+    KEVENT Event;
+    NTSTATUS Status;
+
+    ASSERT(LowerDevice);
+
+    KeInitializeEvent(&Event, NotificationEvent, FALSE);
+    IoCopyCurrentIrpStackLocationToNext(Irp);
+
+    IoSetCompletionRoutine(Irp, ForwardIrpAndWaitCompletion, &Event, TRUE, TRUE, TRUE);
+
+    Status = IoCallDriver(LowerDevice, Irp);
+    if (Status == STATUS_PENDING)
+    {
+        Status = KeWaitForSingleObject(&Event, Suspended, KernelMode, FALSE, NULL);
+        if (NT_SUCCESS(Status))
+            Status = Irp->IoStatus.Status;
+    }
+
+    return Status;
+}
+
+static
+NTSTATUS
+FdoHandleQueryCompatibleId(
+    _Inout_ PIRP Irp)
+{
+    PWCHAR GenScsiAdapterId = L"GEN_SCSIADAPTER";
+    PWCHAR Ids, NewIds;
+    ULONG Length = 0;
+    ULONG i;
+
+    Ids = (PWCHAR)Irp->IoStatus.Information;
+    if (Ids)
+    {
+        /* Calculate the length of existing MULTI_SZ value */
+        i = 0;
+        while (Ids[i] != UNICODE_NULL || Ids[i + 1] != UNICODE_NULL)
```

Comment:
Why not using instead:
```c
Length = 0;
while (*Ids)
{
    Length += wcslen(Ids) + 1;
    Ids += Length;
}
```

Reply:
This one counts each "line" of MULTI_SZ, and could allow getting rid of `i` variable, good idea!